### ISSUE TITLE
RX 5700 XT, Hashrate Bottleneck, and possible cause

### ISSUE THREAD

**Issue by Josevora** [ethereum-mining/ethminer#1917]
**Describe the bug**
I get 200-500KH/s speed on both OpenCL Kernels...

Hi guys again,

so, I've been messing around with the Ethminer Opencl Kernel, and I think I found out what is wrong with the RX 5700 family for mining ETH on OpenCL 2.0:

(Gonna need some help fixing it)

Let's Start:

After you download the new release's source code here:
https://github.com/ethereum-mining/ethminer/archive/v0.19.0-alpha.0.tar.gz
Extract it and find the kernels/cl/ folder,
Edit the ethash.cl file, change this line(236):
 `mix = fnv(mix, g_dag[buffer[lane_idx]].uint8s[thread_id]); \ `
to this:
`   /* mix = fnv(mix, g_dag[buffer[lane_idx]].uint8s[thread_id]);*/ \  `
compile and run the miner!

That line is responsible for bottlenecking the hole code!

I am able to achieve 300+ MH/s without that line of code!
(LOOOOL not mining properly, of course, I get only rejected shares!)

So:

My thoughts on this, and after some testing, the problem is MEMORY ACCESS!
let's get back to the code:
`   mix = fnv(mix, g_dag[buffer[lane_idx]].uint8s[thread_id]); \ `
if I change `buffer[lane_idx]` for a defined value eg. '0',
It runs @ ~80MH/s (I assume It's the real Speed),
but of course again, REJECTED SHARES, because `buffer[lane_idx]` is supposed to change according to `lane_idx`!
Bottom line is, fetching this piece of code's value from memory is slowing down the hole thing!

Conclusion:

I need Help, since I'm no ETH nor OpenCL Expert...
We can assume from this the Theoretical Seed of ~80MH/s for mining ETH on the RX 5700 XT!
(not bad, but I was Expecting 100+ MH/s)
(maybe with specific optimizations, driver upgrade, etc, we could get better performance)

**Comment by Josevora** [comment]
yes, I just copy-pasted my post from Bitcointalk....

**Comment by MariusVanDerWijden** [comment]
Fetching the piece from memory is exactly what was used to prevent the creation of ASICs. It's part of the algorithm and can not be optimized away.

**Comment by AndreaLanfranchi** [comment]
@Josevora 
as @MariusVanDerWijden pointed out you can't remove that memory access path.

Nevertheless your depicted hashrate is way too low (200 to 500 Kh).
You should be *at least* in the 30**Mh** range

**Comment by Josevora** [comment]
@MariusVanDerWijden, you missed my point, and @AndreaLanfranchi got it right...
the problem is not fetching the data... the problem is HOW LONG its taking to do that... please take this into consideration:
When I use a defined value for `buffer`, the will make it ALWAYS fetch the same data from `g_dag`... It's fetching data and it's giving me now 65MH/s... 
The problem is fetching that value from `buffer` that is declared here(279): `__local uint buffer[WORKSIZE];` ! I will try to mess around with this declaration... 
That is something that works fine on other cards, but takes way to long on this card...

The optimization that we need is not "deleting that line", but adapt it so the XT would work.....

**Comment by BoriZka67** [comment]
@Josevora
Hello. I'm also testing RX5700XT. I will be glad if You succeed in your plan. Please share the result. Thank you for your hard work!

**Comment by Josevora** [comment]
@BoriZka67 ,
this has been fixed... By AMD should I say...
The issue was with the driver itself, was not allowing for normal communication between core and memory!
All good now!
